middle_end/flambda2/to_cmm/to_cmm_expr.ml

```diff
@@ -0,0 +1,825 @@
+  in
+  let make_arm ~tag_discriminant env res (d, action) =
+    let d = prepare_discriminant ~tag:tag_discriminant d in
+    let cmm_action, res = apply_cont env res action in
+    (d, cmm_action), res
+  in
+  match Targetint_31_63.Map.cardinal arms with
+  (* Binary case: if-then-else *)
+  | 2 -> (
+    let aux = make_arm ~tag_discriminant env in
+    let first_arm, res = aux res (Targetint_31_63.Map.min_binding arms) in
+    let second_arm, res = aux res (Targetint_31_63.Map.max_binding arms) in
+    match first_arm, second_arm with
+    (* These switchs are actually if-then-elses. On such switches,
+       transl_switch_clambda will introduce a let-binding to the scrutinee
+       before creating an if-then-else, introducing an indirection that might
+       prevent some optimizations performed by selectgen/emit when the condition
+       is inlined in the if-then-else. *)
+    | (0, else_), (_, then_) | (_, then_), (0, else_) ->
+      wrap (C.ite scrutinee ~then_ ~else_), res
+    (* Similar case to the if/then/else but none of the arms match 0, so we have
+       to generate an equality test, and make sure it is inside the condition to
+       ensure selectgen and emit can take advantage of it. *)
+    | (x, if_x), (_, if_not) ->
+      wrap (C.ite (C.eq (C.int x) scrutinee) ~then_:if_x ~else_:if_not), res)
+  (* General case *)
+  | n ->
+    (* The transl_switch_clambda expects an index array such that index.(d) is
+       the index in [cases] of the expression to execute when [e] matches
+       [d]. *)
+    let max_d, _ = Targetint_31_63.Map.max_binding arms in
+    let m = prepare_discriminant ~tag:tag_discriminant max_d in
+    let unreachable, res = C.invalid res ~message:"unreachable switch case" in
+    let cases = Array.make (n + 1) unreachable in
+    let index = Array.make (m + 1) n in
+    let _, res =
+      Targetint_31_63.Map.fold
+        (fun discriminant action (i, res) ->
+          let (d, cmm_action), res =
+            make_arm ~tag_discriminant env res (discriminant, action)
+          in
+          cases.(i) <- cmm_action;
+          index.(d) <- i;
+          i + 1, res)
+        arms (0, res)
+    in
+    (* CR-someday poechsel: Put a more precise value kind here *)
+    (* The cases of the switch must have a value kind (i.e. they cannot be
+       unboxed). This is because each case will either end in a `Cexit` or end
+       by "calling" the return continuation, in which case it will just return
+       the argument to that continuation. Currently functions cannot return
+       unboxed values, so that argument must have a value kind. *)
+    ( wrap
+        (C.transl_switch_clambda Debuginfo.none (Vval Pgenval) scrutinee index
+           cases),
+      res )
+
+(* CR gbury: for the future, try and rearrange the generated cmm code to move
+   assignments closer to the variable definitions Or better: add traps to the
+   env to insert assignemnts after the variable definitions. *)
```

Comment:
I'm not sure whether this CR is still relevant. From what I recall, it was because originally the initialization code for the modules would end up looking like that:
```
let dummy_module_block = <initilization computation> in
let f1 = read_field dummy_module_block 1 in
let f2 = read_field dummy_module_block 2 in
...
write_field actual_module_block 1 f1;
write_field actual_module_block 2 f2;
...
```
Which would make values `f1`, `f2` and so on be live way longer than necessary.

However, I'm not sure we are still in that case, and things on the flambda2 simplifier side may have changed enough that we don't end up in that kind of situation often. So, we might need to check what happens and potentially remove this comment.

Reply:
Ah, this doesn't happen any more.  I'll remove the CR.